Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+    "def"
+    OneOrMore(.whitespaceOrNewline)
+}
+```
+
+### Control characters: `\t`, `\r`, `\n`, `\f`, `\0`, `\e`, `\a`, `\b`, `\cX`
+
+These escaped literal characters represent specific control characters only.
+
+- `\t`: `CHARACTER TABULATION` (U+0009)
+- `\r`: `CARRIAGE RETURN (CR)` (U+000D)
+- `\n`: `LINE FEED (LF)` (U+000A)
+- `\f`: `FORM FEED (FF)` (U+000C)
+- `\0`: `NUL` (U+0000)
+- `\e`: `ESCAPE` (U+001B)
+- `\a`: `BELL` (U+0007)
+- `\b`: `BACKSPACE` (U+0008) (within a character set only)
+- `\cX`: The control character indicated by Control-`X`; in the code point range `0..<32`.
+
+**DSL:** The corresponding patterns are available as named patterns, or alternatively as character literals.
+
+- `\t`: `.tabCharacter` or `"\u{09}"`
+- `\r`: `.carriageReturn` or `"\u{0d}"`
+- `\n`: `.lineFeed` or `"\u{0a}"`
+- `\f`: `.formFeed` or `"\u{0c}"`
+- `\0`: `.nul` or `"\u{0}"`
+- `\e`: `.escape` or `"\u{1b}"`
+- `\a`: `.bell` or `"\u{07}"`
+- `\b`: `.backspace` or `"\u{08}"`
+- `\cX`: `.control(_: UnicodeScalar)`
+
+
+### Unicode values: `\u`, `\U`, `\x`
+
+Metacharacters that begin with `\u`, `\U`, or `\x` match a character with the specified Unicode scalar values. The format used by Swift string literals (`\u{hhh...}`) and several other legacy regular expression formats are all supported.
+
+- `\u{hhh...}` *(preferred)*: Can use 1 or more hexidecimal digits.
+- `\xhh`: Must use 2 hexidecimal digits.
+- `\uhhhh`: Must use 4 hexidecimal digits.
+- `\x{hhh...}`: Can use 1 or more hexidecimal digits.
+- `\Uhhhhhhhh`: Must use 8 hexidecimal digits.
+
+**DSL:** This functionality is supported through the DSL's use of string literals.
+
+```swift
+Pattern {
+    "abc"
+    "\u{032}"
+    "def"
+}
+```
+
+### Unicode named values and properties: `\N`, `\p`, `\P`
+
+`\N{NAME}` matches a Unicode scalar value with the specified name. For example, `/\p{name=HANGUL SYLLABLE GAG}/` matches the character `"각"`.
+
+`\p{PROPERTY}` and `\p{PROPERTY=VALUE}` match a Unicode scalar value with the given Unicode property (and value, if given). Binary properties (such as `White_Space`) and either full or abbreviated general categories (e.g. `Uppercase_Letter` or `Lu`), can be used without a value. For other properties, provide the property and value to match. For example, `/\p{Numeric_Type=Decimal}+/` matches the full range of the string `"0① ₂³"`.
+
+While most Unicode-defined properties can only match at the Unicode scalar level, some are defined to match an extended grapheme cluster. For example, `/\p{RGI_Emoji_Flag_Sequence}/` will match any flag emoji character, which are composed of two Unicode scalar values.
+
+`\P{...}` matches the inverse of the characters matched by the corresponding `\p{...}` character class.
+
+**DSL:** This functionality is supported through APIs that take a predicate as a parameter, focused either on a `Character` or `UnicodeScalar` instance.
+
+```swift
+Pattern {
+    "abc"
+    OneOrMore(.unicodeProperty(where: \.isAlphabetic))
```

Comment:
I wonder if we'll want a spelling of this that's more amenable to hard-coding in the innards of the matching engine?